ARROW-18247: [JS] fix: RangeError crash in Vector.toArray()

sarfata opened the pull request:
This fixes a bug which happens when a Vector has been created with multiple data blobs and at least one of them has been padded.

See https://observablehq.com/d/14488c116b338560 for a reproduction of the error and more details.

domoritz reviewed (approved):
Thank you for the pull request. Please add a unit test so we make sure this issue doesn't surface again.

sarfata commented:
@domoritz @trxcllnt Thanks for your reviews.

I have added a unit test for my scenario. I also realized that my fix did not work when stride is not 1 so I have added a test (and fix) for that too.

I believe one of you will need to approve running the test workflow again before merging.

trxcllnt commented on js/test/unit/vector/vector-tests.ts:
```diff
@@ -211,6 +211,34 @@ describe(`ListVector`, () => {
 
+describe(`toArray()`, () => {
+    test(`when some data blobs have been padded`, () => {
+        const d1 = vectorFromArray([...Array(16).keys()]);
+        const d2 = vectorFromArray([...Array(10).keys()]);
```
```suggestion
        const d1 = vectorFromArray([...new Array(16).keys()]);
        const d2 = vectorFromArray([...new Array(10).keys()]);
```

trxcllnt commented on js/test/unit/vector/vector-tests.ts:
```diff
@@ -211,6 +211,34 @@ describe(`ListVector`, () => {
+        const d2 = vectorFromArray([...Array(10).keys()]);
+
+        // Padding has been added
+        expect(d2.length < d2.data.length)
```
```suggestion
        expect(d2.length < d2.data.length);
```

domoritz commented:
Looks like a test is failing. 

<img width="656" alt="Screenshot 2022-12-09 at 17 21 24" src="https://user-images.githubusercontent.com/589034/206805101-1f774fc9-c8d3-4e22-a6fe-008394c4dbfd.png">

sarfata commented:
@domoritz fixed. 

🤦 I did some last minute cleanup and added this test to show exactly the key condition for this bug but ended up with a failing test. Sorry about this.
